### PR TITLE
Check Allowance before doing approvals

### PR DESCRIPTION
**Pull request by FrederikBolding**
Skips multi tx's which are unnecessary approvals for ERC20's.

### CODE REVIEW

**Comment by FrederikBolding** [comment]
[ch6865]

File: src/utils/useTxMulti/actions.ts [hunk 1]
```diff
@@ -21,17 +30,31 @@ export const init = (dispatch: Dispatch<TxMultiAction>) => async (
...
   try {
     const txs = await getTxs();
+    const filteredTxs = await filterAsync(txs, async (tx) => {
+      if (network && tx.type === ITxType.APPROVAL && tx.from) {
+        try {
+          const { _spender, _value } = ERC20.approve.decodeInput(tx.data);
+          const provider = new ProviderHandler(network);
+          const allowance = await provider.getTokenAllowance(tx.to, tx.from, _spender);
+          const hasAllowance = bigify(allowance).gte(bigify(_value));
+          return !hasAllowance;
```
Comment (vilsbole) [inline src/utils/useTxMulti/actions.ts:47]: unecessary variable

File: src/utils/useTxMulti/actions.ts [hunk 2]
```diff
@@ -21,17 +30,31 @@ export const init = (dispatch: Dispatch<TxMultiAction>) => async (
...
   dispatch({ type: ActionTypes.INIT_REQUEST });
   try {
     const txs = await getTxs();
+    const filteredTxs = await filterAsync(txs, async (tx) => {
+      if (network && tx.type === ITxType.APPROVAL && tx.from) {
+        try {
+          const { _spender, _value } = ERC20.approve.decodeInput(tx.data);
+          const provider = new ProviderHandler(network);
+          const allowance = await provider.getTokenAllowance(tx.to, tx.from, _spender);
+          const hasAllowance = bigify(allowance).gte(bigify(_value));
```
Comment (vilsbole) [inline src/utils/useTxMulti/actions.ts:46]: invert comparison

File: src/utils/asyncFilter.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export const mapAsync = <T>(array: T[], fn: (arg: T) => Promise<any>) => {
+  return Promise.all(array.map(fn));
+};
+export const filterAsync = <T>(array: T[], fn: (arg: T) => Promise<boolean>) => {
+  return mapAsync(array, fn).then((_arr) => array.filter((_v, i) => !!_arr[i]));
+};
```
Comment (vilsbole) [inline src/utils/asyncFilter.ts:6]: Test Please

**Review by vilsbole — CHANGES_REQUESTED** [review]

**Review by vilsbole — APPROVED** [review]

**Pull request merged**